src/pip/_internal/index.py

```diff
@@ -1313,6 +1313,9 @@ def _get_encoding_from_headers(headers):
     return None
 
 
+_safe_chars_re = re.compile(r'([/@]|%2F|%40)', re.I)
```

Comment:
Given that you're calling `quote(unquote())` on what's left (and in particular that the return value will wind up quoted anyways), it seems like you shouldn't need to be including the quoted variants in the regex. Really, it should just be splitting on the individual characters you want to preserve as is. You can use `safe=''` to be sure that `/` gets quoted in the quote direction.

Reply:
(Given the choice, it's better to let `urllib.parse` do its job rather than trying to do some of it "by hand." This was part of the motivation for removing the use of regexes in previous PR's that touched `_clean_link()`.)

Reply:
IIUC if the quoted variants are not preserved then after the initial unquote it would not be possible to distinguish a `/`/`@` that was intended to be used as-is (to separate path parts or to separate the main URL from the vcs version) from one that was quoted as part of the credentials.

Reply:
The quoted and unquoted variants will already be distinguished because that's what the regex will be doing (separating them into two groups). The only ones getting unquoted will be the ones that should be requoted.

Reply:
@cjerdonek 

> Really, it should just be splitting on the individual characters you want to preserve as is. You can use `safe=''` to be sure that `/` gets quoted in the quote direction.

You are right! It does work.

The only downside is it makes %xx escaping uppercase, for example:

```
>>> from urllib.parse import quote, unquote
>>> print(quote(unquote("2F%2f"), safe=""))
2F%2F
```

Not sure whether the case-sensitivity is important in that case.

Reply:
Thanks for seeing the point.

> The only downside is it makes %xx escaping uppercase, for example:

I would consider the normalization a feature, actually. Also, `_clean_link()` is already normalizing other escape sequences in that way, so I think having all escape sequences treated consistently is another benefit.

Reply:
Better than me, I skipped right over `skip=''`. :)

Reply:
Suddenly i've figured out that it should be split on `re.compile(r'(@|%2F', re.I)` and it seems work now!

Reply:
Are you sure that any special characters need to be preserved in the "path" case? Can you provide an example?

Reply:
It looks like the "path" case only '@' needs to be preserved. For the URL paths -- '@' and '%2f'.

Reply:
Thanks, @atugushev. Yes, I suspected they are different. (I also had some other edge cases to share, but I no longer need to share those.)

Because the logic is different in at least a couple respects, I think the "path" and "url" cases should be handled by different code, as opposed to e.g. trying to use the same splitting function for both. I would also recommend doing the path case in a separate PR, because there is some added trickiness that would be worth discussing separately. And if the URL case is going to be handled separately from the path case, that would let you go back to leaving the quoted variants out of the regex, as I suggested earlier above.